tutorials/14-minituring/14_minituring.jmd

```diff
@@ -0,0 +1,327 @@
+Thus, readers are advised to refer to Turing's documentations and code if knowledge of Turing's context system is required.
+Also, it should be pointed out that the context system design in Turing is not in its definitive form as of the current version at the time of this notebook is written.*
+
+```julia
+abstract type AbstractContext end
+```
+
+```julia
+struct SamplingContext{R} <: AbstractContext where {R<:Random.AbstractRNG}
+    rng::R
+end
+
+SamplingContext() = SamplingContext(Random.default_rng())
+
+function observe(context::SamplingContext, varinfo, sampler, dist, var_id, var_value)
+    logp = Distributions.loglikelihood(dist, var_value)
+    varinfo[var_id] = (var_value, logp)
+    return var_value # return value not used, only for coherent behavior as `assume`
+end
+
+function assume(context::SamplingContext, varinfo, sampler, dist, var_id)
+    sample = Random.rand(context.rng, dist)
+    logp = Distributions.loglikelihood(dist, sample)
+    varinfo[var_id] = (sample, logp)
+    return sample
+end;
+```
+
+Then, we'll encounter our first nontrivial piece of Julia code, that is the compiler.
+In one sentence, the compiler will replace every tilde notation with a call to `observe` or `assume` function defined in the `context` code block.
+
+> **Julia Sidebar**
+> I am afraid there is not easy way to explain what happens here. So I am going leave some references, and once the reader go though the referenced material, the following block should not be too difficult to understand.
+> 
+>   - [Julia Metaprogramming](https://docs.julialang.org/en/v1/manual/metaprogramming/)
+>   - [MacroTools](https://fluxml.ai/MacroTools.jl/dev/pattern-matching/)
+>   - [Turing Compiler Design](https://turing.ml/dev/docs/for-developers/compiler) (The implementation here is heavily simplified compared to Turing's compiler.)
+
+```julia
+macro mini_model(expr)
+    return esc(mini_model(expr))
+end
+
+function mini_model(expr)
+    def = MacroTools.splitdef(expr)
+
+    # `MacroTools.postwalk` is a utility function for traverse AST in post-order
+    def[:body] = MacroTools.postwalk(def[:body]) do sub_expr
+        # `MacroTools.@capture` provide a utility function for pattern matching
+        if MacroTools.@capture(sub_expr, var_ ~ dist_)
+            if var in def[:args]
+                return quote
+                    $(observe)(context, varinfo, sampler, $dist, $(Meta.quot(var)), $var)
+                end
+            else
+                return quote
+                    $var = $(assume)(context, varinfo, sampler, $dist, $(Meta.quot(var)))
+                end
+            end
+        else
+            return sub_expr
+        end
+    end
+
+    def[:args] = vcat(
+        [
+            # Insert extra arguments
+            :(model::($MiniModel)),
+            :(varinfo::($VarInfo)),
+            :(context::($AbstractContext)),
+            :(sampler::($(AbstractMCMC.AbstractSampler))),
```

Comment:
`sampler` is not a part of `MiniModel` now, still needed in `assume` functions for `MHTransitionContext`, so pass in as an argument.